### PR TITLE
refactor: compliance with the knowledge panels API

### PR DESCRIPTION
**Pull request by sumit-158**
### What
- Compliance with the Knowledge panel API

### Fixes bug(s)
- #39

### CODE REVIEW

File: app/knowledge_panels.py [hunk 1]
```diff
@@ -210,24 +213,28 @@ async def wikidata_kp(self):
...
-                            "wikipedia": val.wikipedia_relation,
-                            "image_url": val.image_url,
-                            "open_street_map": val.OSM_relation,
-                            "INAO": val.INAO_relation,
-                        },
-                    ],
+                    "element_type": "text",
+                    "text_element": {
+                        "source_label": val.label_tag,
+                        "source_description": val.description_tag,
```
Comment (alexgarel) [inline app/knowledge_panels.py:219]: source_label and source_description do not seem to exist

Also you must have a "html" in your text_element !

So:
```suggestion
                        "html": f"<p><em>{val.label_tag}</em></p><p><small>{val.description_tag}</small></p>",
```

File: app/knowledge_panels.py [hunk 2]
```diff
@@ -210,24 +213,28 @@ async def wikidata_kp(self):
...
             )
+
+            html.append(
+                {
+                    "element_type": "links",
+                    "link_element": {
+                        "wikipedia": val.wikipedia_relation,
+                        "image_url": val.image_url,
+                        "open_street_map": val.OSM_relation,
+                        "INAO": val.INAO_relation,
```
Comment (alexgarel) [inline app/knowledge_panels.py:233]: Here you have to build a text_element with a nice html

**Review by alexgarel — COMMENTED** [review]
@sumit-158 you are on the right track :-)

File: app/knowledge_panels.py (hunk 2 above)
Comment (sumit-158) [reply app/knowledge_panels.py]: Made the changes @alexgarel

**Comment by sumit-158** [comment]
Made the changes @alexgarel

File: app/knowledge_panels.py [hunk 3]
```diff
@@ -225,12 +224,16 @@ async def wikidata_kp(self):
...
-                        "open_street_map": val.OSM_relation,
-                        "INAO": val.INAO_relation,
+                    "element_type": "text",
+                    "text_element": {
+                        "html": (
+                            "<ul>"
+                            f"<li><a href={val.wikipedia_relation}'>wikipedia</a></li>"
+                            f"<li><a href={val.image_url}>wikidata thumbnail</a></li>"
+                            f"<li><a href={val.OSM_relation}>OpenSteetMap Relation</a></li>"
+                            f"<li><a href={val.INAO_relation}>INAO relation</a></li>"
```
Comment (alexgarel) [inline app/knowledge_panels.py:234]: There are different problem:
1. the first link lacks opening quote
2. image_url is a image url so we should display it as an image instead :-) (using )
3. you forgot to add quotes around urls 
4. for OpenStreet map relation and INAO you should condition it by the fact that they should not be none or ""
eg. for http://localhost/knowledge_panel?lang_code=fr&facet_tag=category&value_tag=en:beers

I think it's worth building a list of string and using "".join here.

File: app/models.py [hunk 4]
```diff
@@ -105,100 +105,60 @@ def facet_plural(facet: str):
...
-WikidataPanel = Union[TextFacetWikiData, WikiDataLinksItem]
 
 
-class WikidtaElement(BaseModel):
-    """
-    contains wikidata description and text elements
-    """
+class KnowledgePanelItem(BaseModel):
+    # Helper class for reccuring item
+    elements: Optional[list] = None
```
Comment (alexgarel) [inline app/models.py:145]: Here you can immediately precise it's a list of BaseElement

```suggestion
    elements: Optional[list[BaseElement]] = None
```

File: app/models.py [hunk 5]
```diff
@@ -105,100 +105,60 @@ def facet_plural(facet: str):
...
-    title: str
-    elements: list
+    elements: Optional[list[BaseElement]] = None
 
 
-class WikidataKnowledgePanelItem(BasePanel):
-    elements: Optional[list[WikidtaElement]] = None
+class WikidataKnowledgePanelItem(BaseModel):
+    elements: Optional[list[BaseElement]] = None
+    title_element: BaseTitleElement
```
Comment (alexgarel) [inline app/models.py:157]:
```suggestion
class WikidataKnowledgePanelItem(KnowledgePanelItem):
```

File: app/models.py [hunk 6]
```diff
@@ -105,100 +105,60 @@ def facet_plural(facet: str):
...
 
 
-class HungerGameKnowledgePanelItem(BasePanel):
-    elements: Optional[list[HungerGameElement]] = None
+class HungerGameKnowledgePanelItem(KnowledgePanelItem):
+    elements: Optional[list[BaseElement]] = None
```
Comment (alexgarel) [inline app/models.py:161]: you don't need this, it does not add anything.
```suggestion
```

File: app/knowledge_panels.py (hunk 3 above)
Comment (sumit-158) [reply app/knowledge_panels.py:234]: indeed!

**Review by alexgarel — APPROVED** [review]
Nice ! Let's merge :-)

**Pull request merged**